### DISCUSSION TITLE
SSL Certificate "no start line"

### DISCUSSION THREAD

**Discussion by strawberryfly** [outline/outline#3491] [Self Hosting] [answered]
Hello everyone,

I have a problem with the new SSL without proxy setting. We have our own CA and have created a certifcate for outline. I have added the paths to the .env, however on every startup the following error message comes up:

`Error: error:0909006C:PEM routines:get_name:no start line`
    `at node:internal/tls/secure-context:65:13`
    `at Array.forEach (<anonymous>)`
    `at setCerts (node:internal/tls/secure-context:63:3)`
    `at configSecureContext (node:internal/tls/secure-context:152:5)`
    `at Object.createSecureContext (node:_tls_common:117:3)`
    `at Server.setSecureContext (node:_tls_wrap:1344:27)`
    `at Server (node:_tls_wrap:1203:8)`
    `at new Server (node:https:69:3)`
    `at Object.createServer (node:https:105:10)`
    `at start (/opt/outline/build/server/index.js:84:68)`
    `at throng (/opt/outline/node_modules/throng/lib/throng.js:24:18)`
    `at Object.<anonymous> (/opt/outline/build/server/index.js:131:21)`
    `at Module._compile (node:internal/modules/cjs/loader:1103:14)`
    `at Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)`
    `at Module.load (node:internal/modules/cjs/loader:981:32)`
    `at Function.Module._load (node:internal/modules/cjs/loader:822:12)`
    `at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)`
    `at node:internal/main/run_main_module:17:47 {`
  `library: 'PEM routines',`
  `function: 'get_name',`
  `reason: 'no start line',`
  `code: 'ERR_OSSL_PEM_NO_START_LINE'`
`}`

The structure of the Certificates
Private.key
`-----BEGIN RSA PRIVATE KEY-----`
`BASE64 Key`
`-----END RSA PRIVATE KEY-----`

Public.pem (Cert, intermediate, RootCa)
`-----BEGIN CERTIFICATE-----`
`BASE64 Key`
`-----END CERTIFICATE-----`

The keys are in Base64 format. I have also "checked" the certificate using openssl. Does anyone have a similar problem or knows what the problem could be ?

**Comment by tommoor** [comment]
The entire thing should be Base64 encoded including the first and last lines, could this be it?

**Comment by strawberryfly** [comment]
i have now encoded the entire keys with base64, however i get the same error message :(

**Reply by tommoor** [reply]
If you can post the value that you're putting in the env for the public cert here then that might offer some clues

**Comment by almereyda** [comment]
When searching the Web for `PEM routines:get_name:no start line` or `ERR_OSSL_PEM_NO_START_LINE` I'm presented with a choice of possible sources and solutions for the problem.

https://stackoverflow.com/questions/63030755/error-error0909006cpem-routinesget-nameno-start-line-node suggests it could be something with file encodings, or some `.npmrc` configuration that should be deleted (for no given reason).

https://stackoverflow.com/questions/66691292/npm-err-code-err-ossl-pem-no-start-line-and-routinesget-nameno-start-line-whe also goes the rather rough way to delete pieces and reset a Node installation (also without adding reasons why this might help).

https://github.com/auth0/node-jsonwebtoken/issues/642 also has more approaches to why this might fail. A repeating pattern is something around `\n` line breaks. Another example simply swapped public and private key (:

From the pasted paths, I conclude this is a *nix system, or some BSD, following POSIX file system conventions?
Is your CA hosted in a similar environment, or eventually sitting on a Windows system?

**Comment by strawberryfly** [comment, marked as answer]
I have now used the Docker+FileSystem method and it worked directly with the same certificates. Thanks for the help

**Reply by dsampaioneova** [reply]
Can You give me a hand? I still have the same problem

**Reply by strawberryfly** [reply]
i mount the certificate in the docker container with 
`--mount type=bind,src=/root/YourPrivate.pem,dst=/opt/outline/private.pem --mount type=bind,src=/root/YourPublicCA.pem,dst=/opt/outline/public.pem`

and i set in the .env
`#Extra variablen
NODE_EXTRA_CA_CERTS=/opt/outline/public.pem`

Hope this helps :)

**Comment by ciekawy** [comment]
for anyone facing similar issue and looking here - it happened to me that my certs got formatted in one line eg `-----BEGIN CERTIFICATE----- MIyMIFxDxUn8 ...` - while this should work, some libraries are expecting a new line after `-----BEGIN CERTIFICATE----- ` and before `-----END CERTIFICATE-----`. I solved my issue after making sure the certs/keys are provided with mentioned new lines

**Reply by SimonVanMello** [reply]
This worked for me

**Comment by sixpolys** [comment]
For me it worked to comment out the cert and key variable in the .env file

**Comment by gen0v** [comment]
For me it worked like this:

1) Comment variables inside docker.env out
```
# inside docker.env comment lines out like this:
...
#SSL_KEY=
#SSL_CERT=
...
```
2) Create base64 strings for cert/key
```
# cat cert.pem | base64
# cat key.pem | base64
```
3) Put as enviroment variable inside compose file like this:
```
...
services:
  outline:
    image: docker.getoutline.com/outlinewiki/outline:latest
    env_file: ./docker.env
	...
    environment:
      SSL_CERT: |
        <base64 string
		...
        ...>
      SSL_KEY: |
        <base64 string
		...
		...>
...
```
